fabric-screen-api-v1/src/main/java/net/fabricmc/fabric/impl/client/screen/ButtonList.java

```diff
@@ -32,56 +33,63 @@
 @ApiStatus.Internal
 @Environment(EnvType.CLIENT)
 public final class ButtonList extends AbstractList<ClickableWidget> {
-	private final List<Selectable> listeners;
+	private final List<Drawable> drawables;
+	private final List<Selectable> selectables;
 	private final List<Element> children;
 
-	public ButtonList(List<Selectable> listeners, List<Element> children) {
-		this.listeners = listeners;
+	public ButtonList(List<Drawable> drawables, List<Selectable> selectables, List<Element> children) {
+		this.drawables = drawables;
+		this.selectables = selectables;
 		this.children = children;
 	}
 
 	@Override
 	public ClickableWidget get(int index) {
-		return (ClickableWidget) listeners.get(translateIndex(index, false));
+		final int drawableIndex = translateIndex(drawables, index, false);
+		return (ClickableWidget) drawables.get(drawableIndex);
 	}
 
 	@Override
 	public ClickableWidget set(int index, ClickableWidget element) {
-		index = translateIndex(index, false);
-		int prevIndex = listeners.indexOf(element);
+		final int drawableIndex = translateIndex(drawables, index, false);
+		drawables.set(drawableIndex, element);
 
-		if (prevIndex >= 0 && prevIndex != index) {
-			if (prevIndex < index) index--;
-			listeners.remove(prevIndex);
-		}
-
-		int childIndex = children.indexOf(element);
+		final int selectableIndex = translateIndex(selectables, index, false);
+		selectables.set(selectableIndex, element);
 
-		if (childIndex >= 0) {
-			children.set(childIndex, element);
-		}
-
-		return (ClickableWidget) listeners.set(index, element);
+		final int childIndex = translateIndex(children, index, false);
+		return (ClickableWidget) children.set(childIndex, element);
 	}
 
 	@Override
 	public void add(int index, ClickableWidget element) {
-		index = translateIndex(index, true);
-
-		if (listeners.remove(element)) { // ensure no duplicates
+		// ensure no duplicates
+		final int duplicateIndex = drawables.indexOf(element);
+		if (duplicateIndex >= 0) {
+			drawables.remove(element);
+			selectables.remove(element);
 			children.remove(element);
-			index--;
+			if (duplicateIndex <= translateIndex(drawables, index, true)) { 
+				index--;
+			}
 		}
 
-		listeners.add(index, element);
-		this.children.add(element);
+		final int drawableIndex = translateIndex(drawables, index, true);
+		drawables.add(drawableIndex, element);
+
+		final int selectableIndex = translateIndex(selectables, index, true);
+		selectables.add(selectableIndex, element);
+
+		final int childIndex = translateIndex(children, index, true);
+		children.add(childIndex, element);
 	}
 
 	@Override
 	public ClickableWidget remove(int index) {
-		index = translateIndex(index, false);
+		index = translateIndex(drawables, index, false);
```

Comment:
then this translate should happen on children than drawables as we use children as the final reference